### PR TITLE
Fix incorrect restart count for docker containers

### PR DESCRIPTION
**Pull request by yujuhong**

### CODE REVIEW

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -375,7 +375,10 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
 	for _, status := range pod.Status.ContainerStatuses {
 		oldStatuses[status.Name] = status
 		if status.LastTerminationState.Terminated != nil {
-			lastObservedTime[status.Name] = status.LastTerminationState.Terminated.FinishedAt
+			timestamp, ok := lastObservedTime[status.Name]
+			if !ok || (ok && timestamp.Before(status.LastTerminationState.Terminated.FinishedAt)) {
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: nit: The second `ok &&` check isn't necessary right?

Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: What triggers this issue? Why are you introducing this check? I see the correctness, but I want to understand the motivation.

Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: yep. you are right. will remove it here and in line 447.

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -435,20 +438,19 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
...
-					// container name as done.
-					containerDone.Insert(dockerContainerName)
-				}
+			if terminationState == nil {
+				// Not a dead container.
+				continue
 			}
-			if count {
+			// Only count dead containers terminated after last time we observed,
+			lastObservedTime, ok := lastObservedTime[dockerContainerName]
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: This breaks on restarts right?

File: pkg/kubelet/dockertools/manager.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: The bug was there for a while, but was masked by the previous restart count bug, which was fixed by my last PR. We are now seeing really high restart counts due to incorrect accounting.

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: Not really. We get the restart count that we last persisted to the apiserver. In most cases, you should still see monotonically increasing restart counts. (I tried this by killing the kubelet in my cluster). This, however, does break for static pods over retarts.

**Comment by yujuhong** [comment]
@vishh, just a clarification. This PR, is to fix the insane restart counts I saw in my cluster. For the longer term, we'll use the docker label to record the count, as you suggested. I haven't had time to get to that yet, and this part of code is quite messy :(

**Comment by vishh** [comment]
@yujuhong: Thanks for clarifying :) LGTM

**Pull request merged**